### DISCUSSION TITLE
8.5.2 and IKVM.Tools.Importer error "fatal error IKVMC5025: Unable to load runtime assembly"

### DISCUSSION THREAD

**Discussion by marcomoret** [ikvmnet/ikvm#373] [Q&A]
I try to convert a .jar into dll file but, from 8.5.2 I get this discussion subject error
I run this simple command ./ikvmc.exe -out:output.dll input.jar into this folder
...\ikvm.msbuild.tools.runtime.win7-x64\8.5.2\ikvmc\net461\win7-x64\
where is the problem? With old version of ikvm all run correctly

**Comment by wasabii** [comment]
ikvmc requires that you use -reference to point to every DLL from the target framework that you are compiling the assembly for, as well as the -runtime argument to point to the IKVM.Runtime.dll assembly that you want the resulting assembly to bind to.

These options exist because there is no longer a single .NET Framework for which you might be generating an assembly. There are like 200 different versions of Core, at various version levels, that you might use the same tool to operate against, and multiple versions of IKVM.Runtime you might want to cross target (using the .NET Core 3 version of the tool to target a .NET 5 version of the assembly, etc).

The required arguments have thus become very complicated.

Which is why IkvmReference and MavenReference now exist. These let you directly add a Java library from Maven, or from a raw .JAR file, to your .NET project, as part of it's build, where the build scripts assemble the appropriate options based on what your own library targets.

**Reply by marcomoret** [reply]
Can you write me a sample step by step procedure to create my output.dll file?

**Comment by wasabii** [comment]
I would consult the README for the ideal steps.

**Reply by marcomoret** [reply]
thanks

**Comment by marcomoret** [comment]
Maybe I understand the problem but not the solution.
When I use ikmvc.exe 8.5.2 to trasform a jar file into dll I find this error:

IKVM.ByteCode need System.Runtime.CompilerServices.Unsafe, Version=6.0.0.0

but if I put the correct System.Runtime.CompilerServices.Unsafe.dll, I find this other error

System.Memory need System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1 

Maybe the problem is System.Memory.dll included into ikvm 8.5.2, that dll need and old System.Runtime.CompilerServices.Unsafe.dll version

**Reply by marcomoret** [reply]
or the problem is IKVM.ByteCode that use a more recent System.Runtime.CompilerServices.Unsafe.dll

**Reply by wasabii** [reply]
I do not know how you are building your application, but it sounds like you're copying DLLs around by hand and making a general mess of things. The problem you're talking about with mismatched version is generally, on Framework, a problem with assembly binding redirects. Which are auto generated when building things properly in MSBuild, using proper PackageReferences and IkvmReference and stuff. These are dozens of lines in your final generated app.config file, specifying the unified version of assemblies to use at runtime.

I'm sure you can sort it out by hand. But I'm not going to work through it with you when you haven't tried to follow the recommendations.

I'll say this again: just follow the instructions in the README. Unless there is some very specific, and rare, reason you cannot do that, it'll take care of things for you.

**Reply by marcomoret** [reply]
thanks but the real problem is that my little application use a file.dll trasformed by ikvm 8.4.5 from a file.jar
If I update the ikvm packages from 8.4.5 to 8.5.2 or 8.6.2 every calls to my file.dll goes always in error and is not more possible transform the jar into dll for the previus dll version errors
Sorry for my bad previous description

**Reply by wasabii** [reply]
So you're trying to use a DLL built against one version of IKVM against another version of IKVM? That isn't going to work......

**Reply by marcomoret** [reply]
indeed, but it was only a test because I can't recreate the DLL anymore with IKVM versions 8.5.2 and 8.6.2

**Reply by wasabii** [reply]
Well I feel like I already answered that.

**Reply by marcomoret** [reply]
the readme say this:

Statically: By compiling a Java application into a .NET assembly using <MavenReference>, <IkvmReference> or ikvmc....

but I don't find the ikvmc sample use into readme, can you help me?
thanks

**Reply by wasabii** [reply]
There are no samples. I explained the likely issue in the very first answer I gave you. The two options before ikvmc are strongly recommended.

Please try them first.

**Reply by marcomoret** [reply]
I have an old .Net project that include my file.dll created by ikvmc.exe statically
Up to ikvmc.exe 8.4.2 version I used "ikvmc.exe -out:file.dll file.jar" and all worked correctly
from 8.5.2 the dll that ikvmc.exe search have a problem on System.Runtime.CompilerServices.Unsafe.dll because System.Memory.dll 4.0 need System.Runtime.CompilerServices.Unsafe.dll 4.0 but IKVM.ByteCode 8.5.2.0 need System.Runtime.CompilerServices.Unsafe.dll 6.0.
If I try to use a more new System.Memory.dll I find che ikvmc.exe use IKVM.Tools.Importer.dll 8.5.2 that search System.Memory.dll 4.0

**Reply by wasabii** [reply]
>I have an old .Net project that include my file.dll created by ikvmc.exe statically

So fix it.

**Comment by marcomoret** [comment]
Hi, I try to use the first approach: IkvmReference
but no dll is created after rebuild VS project
I make simple steps:
create project
install IKVM package
add .jar like project reference
add PackageReference and IkvmReference in project file

**Reply by wasabii** [reply]
Is it a SDK style project?

**Reply by marcomoret** [reply]
No, I need to create a SDK project?

**Reply by wasabii** [reply]
Yes.

**Reply by marcomoret** [reply]
Hi, sorry but for me is no possible to use new IKVM version from 8.6.2 to last 8.6.4

I try to describe actual problem:

I create the dll form jar file but, when I try to use it into VS project, I get this error:

REG: DisplayName = System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
 (Fully-specified)
REG: Appbase = file:///C:/MM_SMEUP/sorgentiUtili/C#/ikvm_test/IKVM_test_v8.6.4/IKVM_test/bin/Debug/
REG: PrivatePath iniziale = NULL
Assembly chiamante: System.Text.Json, Version=6.0.0.6, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51.

When I add nuget IKVM package into VS project, I see that it install:
System.Text.Json 6.0.6 and System.Memory 4.5.5
why during the project runtime I get an exception on System.Text.Json 6.0.6 that search System.Memory 4.0.1.1 and not System.Memory 4.5.5 like nuget package addition make?
thanks and sorry but I would like to use last IKVM version
regards

**Reply by wasabii** [reply]
I can't figure out whether you're using IkvmReference reference or not. There would be no manual step of you creating the DLL from the jar, if so.

**Reply by marcomoret** [reply]
I did not understand. I'm using a Visual studio project with IKVM package installed via NuGet. The dll was created from the original .jar file but when I try to use it it gives me the exception I wrote to you. until version 8.4.5 this did not happen

**Reply by wasabii** [reply]
How are you trying to use it?

**Reply by marcomoret** [reply]
always with ikvmc.exe utility (last version 8.6.4) I create dll from jar without error (I try also with SDK project but the dll doesn't is created)
After I put the dll like a project reference and install IKVM NuGet package (or also I copy all IKVM dll from 8.6.4 into my project and reference to it, nut the error is the same)

**Reply by wasabii** [reply]
So still not using IkvmReference. I've got to let you handle this one on your own. You've opted to sort through Framework bindings by hand, and I can't do that for you.

**Reply by marcomoret** [reply]
Now I retry SDK with Ikvm reference. I need to install NuGet package in PackageReference mode or Packages.config mode?

**Reply by marcomoret** [reply]
I get this error now
...\obj\Debug\net6.0\<projectname>.AssemblyInfo.cs' not found.'....

**Reply by wasabii** [reply]
There is no such thing as a package.config mode when using SDK style projects, which you have to use, as I mentioned a month ago.

Are you using SDK-style projects?

**Reply by marcomoret** [reply]
now yes. I re-try your suggest steps

**Reply by marcomoret** [reply]
Now I use SDK-style project and PackageReference to install NuGet IKVM package but the problem is the same: ...\obj\Debug\net6.0<projectname>.AssemblyInfo.cs' not found.'....
Have you some idea to solve it?
